Design/Instrument-2.0/design-notes.md

```diff
@@ -0,0 +1,89 @@
+
+### Naming
+
+All class names in this document are preliminary and just for the sake of discussion.
+Better names are needed in several cases.
+
+
+## Design
+
+### Structure
+
+Algorithms that work with workspaces need to access information about spectra.
+Therefore:
+
+- `SpectrumInfo` should be a member of workspace.
+- `SpectrumInfo` owns `DetectorInfo`.
+- `DetectorInfo` owns `InstrumentTree`.
+
+If no detectors are grouped, there is a 1:1 mapping between spectrum and detector.
+`SpectrumInfo` and `DetectorInfo` should thus share data, to avoid waste of memory and CPU resources.
+This can be facilitated by:
+
+- `GeometryDataArray` is used to store vectors of data, such as L2, twoTheta, position, masking flags, monitor flags, ....
+- `SpectrumInfo` and `DetectorInfo` have a field `std::shared_ptr<GeometryDataArray>`. By default both point to the same instance. If detectors are grouped, `SpectrumInfo` allocates its own `GeometryDataArray`.
+
+### Interface
+
+The interface to `SpectrumInfo` and `DetectorInfo` would be almost identical.
+The most typical examples for public methods include:
+
+```cpp
+bool isMonitor(size_t index);
+bool isMasked(size_t index);
+double L2(size_t index);
+double twoTheta(size_t index);
+V3D position(size_t index);
+```
+
+Furthermore, we need to provide access to lower-level data, e.g., via a method `const Detector &DetectorInfo::detector(size_t index)`.
+For `SpectrumInfo` we could have `const std::vector<const Detector *> detectors(size_t index)`.
+
+Note that although the index in the public interface is the same. They relate to two different indexing mechanisms. For example, on `SpectraInfo::isMonitor(size_t index)` the `index` is the spectrum index, which goes from 0 to the maximum number of spectra. On the other hand `DetectorInfo::isMonitor(size_t index)` the `index` is the detector index. This can be larger than the available number of spectra when mapping is not 1:1. 
+
+### Read access
+
+- Access functions take an index which is used for accessing an underlying `std::vector` of data.
+  - The data layout in `GeoemtryDataArray` is not defined, and could be changed in the future.
+  - To facilitate the free data layout, the interface should be `SpectrumInfo::twoTheta(size_t index)` and *not* along the lines of the current implementation with `Workspace::getDteector(size_t index)::twoTheta()`.
+  - Accessors should probably be declared inline for performance reasons.
+
+### Write access
+
+Write access requires modification of all three data structures, `InstrumentTree`, `DetectorInfo`, and `SpectrumInfo`.
+I (Simon) do not yet have an entirely clear picture of this.
+Notes:
+
+- Do we always go via `InstrumentTree` upon modification? For example, masking a spectrum or a detector could be handled at the `SpectrumInfo` or `DetectorInfo` level. These components could also have a copy-on-write mechanism. We would then still share the same `InstrumentTree` for workspaces that have different masking.
```

Comment:
@SimonHeybrock Interesting point. I'm of the opinion that the masking relates more to the data than the fundamental geometry. I wouldn't see maksing as a property of a geometric detector. So on that basis I think that modification would go directly against the `SpectrumInfo` `DetectorInfo`.